File: near_client/lib/src/socket/message_io.dart
import 'package:getx_near/src/model/message.dart';
import 'package:getx_near/src/screen/message/message_controller.dart';
import 'package:getx_near/src/service/message_extention.dart';
import 'package:getx_near/src/socket/socket_base.dart';

class MessageIO extends SocketBase {
  @override
  NameSpace get nameSpace => NameSpace.message;
  @override
  Map<String, dynamic> get query => {"chatRoomId": chatRoomId};

  final MessageController controller;
  MessageIO(this.controller);

  MessageExtention get extention => controller.extention;
  String get chatRoomId => extention.chatRoomId;

  //受信
  void addNewMessageListner() {
    socket.on("new_message", (msg) {
      final newMessage = Message.fromMap(msg);

      if (!newMessage.isCurrent) {
        extention.updateRead(message: newMessage);
        sendUpdateRead([newMessage.id]);
      }

      controller.messages.insert(0, newMessage);
    });
  }

  void addReadListner() {
    socket.on("read", (value) {
      final uid = value["uid"];
      final ids = value["ids"];
      final List<String> temp = List.castFrom(ids);
      for (var id in temp) {
        controller.readUI(id, uid);
      }
    });
  }

  /// 送信
  void sendNewMessage(Message message) {
    socket.emit("new_message", message.toMap());
  }

  void sendUpdateRead(List<String> messageIds) {
    final data = {"uid": extention.currentUser.id, "ids": messageIds};
    socket.emit("read", data);
  }

  void sendUpdateRecent(List<String> userIds) {
    final Map<String, dynamic> data = {
      "chatRoomId": chatRoomId,
      "userIds": userIds,
    };

    socket.emit("update_recent", data);
  }
}

File: near_server/src/utils/socket/message_io.ts
import socket from 'socket.io';

export function messageSocket(
  messageIO: socket.Namespace,
  recentIO: socket.Namespace
) {
  messageIO.on('connection', (socket) => {
    console.log('Chat SOCKET', socket.id);
    const q = socket.handshake.query.chatRoomId;
    if (!q) return;

    const chatRoomId = q as string;

    messageIO.socketsJoin(chatRoomId);

    socket.on('new_message', (msg) => {
      messageIO.to(chatRoomId).emit('new_message', msg);
    });

    socket.on('read', (ids) => {
      messageIO.to(chatRoomId).emit('read', ids);
    });

    socket.on('update_recent', (data) => {
      const roomIds = data['userIds'];
      /// socket.io V4
      recentIO.in(roomIds).emit('update', data);
    });

    socket.on('disconnect', () => {
      socket.leave(chatRoomId);
      console.log('Chat Disconnrect');
    });
  });

  messageIO.on('connect_error', (err) => {
    console.log(`connect_error due to ${err.message}`);
  });
}
